hazelcast/src/test/java/com/hazelcast/client/cluster/ClientClusterReconnectRetryTest.java

```diff
@@ -0,0 +1,73 @@
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.hazelcast.client.cluster;
+
+import com.hazelcast.client.config.ClientConfig;
+import com.hazelcast.client.impl.clientside.HazelcastClientProxy;
+import com.hazelcast.client.impl.connection.ClientConnectionManager;
+import com.hazelcast.client.test.TestHazelcastFactory;
+import com.hazelcast.core.HazelcastInstance;
+import com.hazelcast.test.HazelcastSerialClassRunner;
+import com.hazelcast.test.HazelcastTestSupport;
+import com.hazelcast.test.annotation.NightlyTest;
+import org.junit.After;
+import org.junit.Test;
+import org.junit.experimental.categories.Category;
+import org.junit.runner.RunWith;
+
+import java.util.concurrent.TimeUnit;
+
+import static org.junit.Assert.assertEquals;
+
+@RunWith(HazelcastSerialClassRunner.class)
+@Category(NightlyTest.class)
+public class ClientClusterReconnectRetryTest extends HazelcastTestSupport {
+    private static final int CLIENT_RECONNECT_SECS = 15;
+    // if the test fails because of the slowness, decrease the assertion period below
+    private static final int DISCONNECTED_CLIENT_ASSERTION_PERIOD_SECS = CLIENT_RECONNECT_SECS - 5;
+    private final TestHazelcastFactory factory = new TestHazelcastFactory();
+
+    @Test
+    public void testClientShouldComplyWithConnectionStrategyConfigWhenAttemptingClusterReconnect() {
+        ClientConfig clientConfig = new ClientConfig();
+        // Set connection retry strategy
+        clientConfig.getConnectionStrategyConfig().getConnectionRetryConfig()
+                .setInitialBackoffMillis((int) TimeUnit.SECONDS.toMillis(CLIENT_RECONNECT_SECS))
+                .setMaxBackoffMillis((int) TimeUnit.SECONDS.toMillis(CLIENT_RECONNECT_SECS));
+        HazelcastInstance member = factory.newHazelcastInstance();
+        HazelcastInstance client = factory.newHazelcastClient(clientConfig);
+        ClientConnectionManager clientConnectionManager = ((HazelcastClientProxy) client).client.getConnectionManager();
+        // assert the initial cluster connection
+        assertTrueEventually(() -> assertEquals(1, clientConnectionManager.getActiveConnections().size()));
+
+        member.shutdown();
+        // wait for the client disconnecting from the cluster
+        assertTrueEventually(() -> assertEquals(0, clientConnectionManager.getActiveConnections().size()));
+        factory.newHazelcastInstance();
+        // assert that the client doesn't reconnect to the cluster in DISCONNECTED_CLIENT_ASSERTION_PERIOD_SECS
+        // which is 5 secs less than CLIENT_RECONNECT_SECS
+        assertTrueAllTheTime(() -> assertEquals(0, clientConnectionManager.getActiveConnections().size()),
+                DISCONNECTED_CLIENT_ASSERTION_PERIOD_SECS);
```

Comment:
There is a window of failure for this logic. 

Imagine 
1. We shutdown the member
2. Closed the connection to it on the client-side. 
3. We will trigger reconnection by submitting a task to an executor. https://github.com/hazelcast/hazelcast/blob/master/hazelcast/src/main/java/com/hazelcast/client/impl/connection/tcp/TcpClientConnectionManager.java#L792
4. We pass the assertion with activeConnections.size() == 0
5. We start a new instance

Now, if the execution of the task we submitted in step 3 is delayed somehow, and we start the reconnection procedure after step 5, we would connect immediately and the assertion with ` assertTrueAllTheTime` would fail